### PR TITLE
fix(common/redis): re-AUTH new pool connections with fresh Entra token

### LINKED ISSUE #3554: Redis Entra Id connection stops working after 24hours
## Expected Behavior

We started using Entra Id authentication to Redis with the dapr version 1.14.1. We are able to establish the connection from the AKS pod to Redis using the managed identity used by the pod. The pod runs for many days without restarting. It should be able to connect the Redis whenever there is a request from client.   

## Actual Behavior
However, after running for about 24hours the pod is failing to connect Redis with the error 'State operation failed: the Dapr endpoint indicated a failure. See InnerException for details. Status(StatusCode="Internal", Detail="fail to get PComSearchCacheProduct from state store statestore: ERR WRONGPASS invalid username-password pair") '

## Steps to Reproduce the Problem

Run the pod that connects Redis with Entra ID authentication for more then 24hours. It will fail with the above mentioned error after about 24hours. 

## Additional information
The AKS pod has been working with the same Redis statestore with the password without any issue. We faced this problem when we tried to implement Entra Id authentication.

### PR DESCRIPTION
**Pull request by paulstadler-mesh**
## Description

Fixes [#3554](https://github.com/dapr/components-contrib/issues/3554), the long-standing report that Entra ID auth against Azure Managed Redis stops working after the initial token's TTL elapses.

When `useEntraID: true`, the redis component snapshotted the initial Entra access token into `redis.UniversalOptions.Password` at component-init time. Every new pool connection opened by go-redis afterwards replayed that static AUTH at connection establishment — but the token has a finite TTL (1–24h depending on credential type and audience). After expiry, any new connection (idle timeout, `maxConnAge`, network blip, demand spike, pool churn) fails AUTH with `WRONGPASS invalid username-password pair`.

[#3632](https://github.com/dapr/components-contrib/pull/3632) fixed the background AUTH ACL refresh goroutine being killed by init context cancellation, but only re-AUTHs *existing* connections; it never updated `Options.Password`, so new connections continued to fail with the original snapshot once it expired. This explains the multiple post-#3632 reports on #3554 from users on 1.15.9, 1.16.1, 1.17.x ("token IS being refreshed in logs but pool reconnects fail").

## What this PR does

Installs an `OnConnect` hook on every variant of the underlying go-redis client (v8/v9, plain/cluster/failover) that:

1. Fetches a fresh token from the cached `*azcore.TokenCredential`.
2. Runs `AUTH ACL <oid> <fresh-token>` on the newly-dialed connection.

`azcore` caches tokens internally until ~5 minutes before expiry, so per-connection `GetToken` calls are inexpensive in steady state.

The static `Password` snapshot is no longer set, so go-redis skips its built-in AUTH during connection init and lets `OnConnect` handle authentication. The OID — used as the Redis ACL username, per the [Microsoft Entra for AMR documentation](https://learn.microsoft.com/en-us/azure/redis/entra-for-authentication) — is parsed once from the initial token at component init and stored on `Settings`; `OnConnect` uses it for every subsequent AUTH.

`StartEntraIDTokenRefreshBackgroundRoutine` is unchanged. It remains the safety net for long-lived connections that survive across token-TTL boundaries without ever being pool-cycled.

## Issue reference

Please reference the issue this PR will close: #3554

## Checklist

- [x] Code compiles correctly (`go vet ./common/component/redis/...`, `go build ./common/component/redis/...`, `go build` against all 5 redis-using components: `state/redis`, `pubsub/redis`, `bindings/redis`, `lock/redis`, `configuration/redis`)
- [x] Existing tests pass (`go test ./common/component/redis/... -count=1` — `ok 1.554s`)
- [x] Created/updated tests (`common/component/redis/entraid_onconnect_test.go` asserts `OnConnect` is wired and `Password` is unset when `UseEntraID` is true, and not wired when false, across all v8/v9 node/cluster/failover variants; plus `EntraIDFetchAuthArgs` coverage)
- [ ] Extended the documentation
- [ ] Provided sample for the feature

### Reproduction

The bug was reproduced empirically against an Azure Managed Redis cluster with a workload-identity-bound managed identity that has the default access policy, using daprd 1.17.2:

1. Deploy a sidecar with a `state.redis` component configured as `useEntraID: true, redisType: cluster, enableTLS: true`.
2. Run a workload that exercises the state store directly via `DaprClient.GetState`/`SaveState` (Dapr Actors-only consumers don't surface this because actor state-store access is lazy).
3. Observe pods running cleanly for ~24 hours, then transitioning to `CrashLoopBackOff` with the daprd sidecar logging `WRONGPASS invalid username-password pair` from `state.redis` calls. The crossover aligns with the initial token's TTL.

I also verified the AUTH contract independently with a manual probe from inside the pod:

- Federated workload-identity → Entra exchange succeeds for scope `https://redis.azure.com/.default` and returns a token with the correct `aud` / `appid` / `oid`.
- Sending that token to AMR as `AUTH <token>` (no username) returns `WRONGPASS`.
- Sending the same token as `AUTH <oid> <token>` returns `+OK`.

This matches the documented AMR auth contract — confirms it's a client-side issue, not an authorization-policy issue.

### After this change

- Built a daprd image from `dapr/dapr` v1.17.2 with `go.mod` pointing at this branch via `replace`, deployed to the same cluster, configured the state-store with `maxConnAge: 5m` to force the pool to recycle every minute (rather than waiting for the natural ambient cadence).
- Pods came up cleanly. Both Redis components (`state.redis` and `pubsub.redis`) loaded successfully.
- Sidecar has been running across many full pool-cycle windows (every 5 min the entire pool is replaced; OnConnect fires per dial). **Zero `WRONGPASS` errors** across this period and zero sidecar restarts.
- Long-soak validation (>24h, past the original cliff) is in progress and will be reported back here.

### Edge cases considered

- v8/v9 cluster mode: each backend node returned by `CLUSTER SHARDS` gets the OnConnect hook applied, so dials to discovered IPs use a fresh token each time, not the original snapshot.
- Failover/sentinel: same — `FailoverOptions.OnConnect` is wired.
- The previously-static `Settings.Username` and `Settings.Password` are now empty when `useEntraID` is true. The pre-existing validation in `InitEntraIDCredential` (formerly `GetEntraIDCredentialAndSetInitialTokenAsPassword`) already errored when these were configured externally with `useEntraID`, so no behavior change for misconfiguration.
- Backward compat: components with explicit `redisUsername` / `redisPassword` and `useEntraID: false` are unaffected (the OnConnect hook is gated on `s.UseEntraID`).

### Also fixed here: `AuthACL` pipeline was never executed

While debugging, I noticed that `(c v8Client).AuthACL` and the v9 equivalent used `c.client.Pipeline()` to send the AUTH ACL command, but never called `Pipeline.Exec(ctx)`. In go-redis v8/v9 that means the queued AUTH command was never actually sent over the wire — the background refresh goroutine had been a silent no-op since #3632 landed, so long-lived idle connections never received the periodic re-AUTH the goroutine was meant to provide.

This is now fixed in this PR (both v8 and v9): the pipeline is `Exec`'d and its error propagated. Note that go-redis's `UniversalClient` only exposes `Cmdable` (not `StatefulCmdable`), so AUTH genuinely has to go through a pipeline here — calling `AuthACL` directly on the client doesn't compile; the missing piece was the `Exec` call.

### Other review follow-ups addressed

- `EntraIDFetchAuthArgs` now returns a clear error when the derived OID username is empty, instead of issuing `AUTH ACL "" <token>` and surfacing a confusing `WRONGPASS`.
- Added unit tests (see the checklist above).

## Refs

- #3554 (issue this PR closes)
- #3632 (the prior partial fix this PR completes)
- [Microsoft Learn: Use Microsoft Entra for cache authentication with Azure Managed Redis](https://learn.microsoft.com/en-us/azure/redis/entra-for-authentication)

### CODE REVIEW

**Comment by paulstadler-mesh** [comment]
Hi @yaron2 @shivamkm07 — this PR closes the loop on #3632: that one fixed the AUTH-refresh goroutine, but Options.Password was still snapshotted at init, so new pool connections (idle timeout, max-conn-age, network blip) kept replaying the original token and failing AUTH after its TTL. This wires an OnConnect hook so every dial AUTHs with a freshly-fetched token. Reproduced against Azure Managed Redis, soaked successfully across many maxConnAge: 5m pool-cycle windows.

Could one of you /ok-to-test and take a look? Happy to add a unit test asserting OnConnect is wired when UseEntraID is true if you'll tell me the shape you'd prefer (matching an existing pattern in the package).

**Comment by paulstadler-mesh** [comment]
Friendly bump on this — it's been a few weeks since I tagged @yaron2 / @shivamkm07, and I've rebased onto current `main`.

Some context on impact, since I think it's broader than it might look. This fixes #3554: Entra ID auth against Azure Managed Redis fails with `WRONGPASS` once the initial token's TTL elapses and go-redis opens new pool connections. The merged background-refresh (#3632) keeps the cached credential fresh, but it doesn't re-AUTH connections created *after* a pool recycle — so the failure persists. This PR installs an `OnConnect` hook so every new pool connection authenticates with a freshly-minted token.

We've been running this patch in dev and production against Azure Managed Redis and it cleanly resolves the post-TTL failures.

One thing worth flagging for any Azure/Microsoft folks watching: per the AKS Dapr extension's [issue-handling policy](https://learn.microsoft.com/azure/aks/dapr-overview#issue-handling), Dapr component fixes are expected to land in Dapr OSS first and are then picked up by the managed `-msft` daprd builds. So this PR is effectively the upstream gate for *every* AKS Dapr extension user hitting #3554 on Azure Managed Redis with Entra ID — not just us.

Happy to add tests, split the change, or adjust the approach if that helps it land. Any chance it could target the next 1.17 patch?

File: common/component/redis/redis.go [hunk 1]
```diff
@@ -293,45 +292,69 @@ func StartEntraIDTokenRefreshBackgroundRoutine(client RedisClient, username stri
...
 	}
-	return &token.ExpiresOn, &cred, nil
+	objectID, ok := objectIDRaw.(string)
+	if !ok {
+		return nil, errors.New("redis client configuration error: object ID claim is not a string")
+	}
+
+	s.entraIDUsername = objectID
+	s.entraIDTokenCredential = cred
+	return &token.ExpiresOn, nil
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/redis.go:377]: Good call — fixed in daa94608. `InitEntraIDCredential` now returns a config error when the `oid` claim is present but empty, so a misconfigured token fails fast at init rather than as a confusing connect-time `WRONGPASS`.

File: common/component/redis/v8client.go [hunk 2]
```diff
@@ -318,11 +318,39 @@ func (c v8Client) TTLResult(ctx context.Context, key string) (time.Duration, err
...
+	// UniversalClient only exposes Cmdable (not StatefulCmdable), so AUTH must be
+	// issued via a pipeline. The pipeline must be Exec'd for the command to actually
+	// be sent over the wire — without it the AUTH is silently queued and dropped.
 	pipeline := c.client.Pipeline()
 	statusCmd := pipeline.AuthACL(ctx, username, password)
+	if _, err := pipeline.Exec(ctx); err != nil {
+		return err
+	}
 	return statusCmd.Err()
 }
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/v8client.go:327]: This one doesn't apply: go-redis's `Pipeliner` interface has no `Close()` method (only `Do`/`Process`/`Discard`/`Exec`/`Len`), so `defer pipeline.Close()` would not compile. A plain (non-Tx) pipeline also doesn't hold a dedicated connection — it borrows one from the pool only for the duration of `Exec` and returns it immediately — so there's nothing to leak here.

File: common/component/redis/v9client.go [hunk 3]
```diff
@@ -318,11 +318,39 @@ func (c v9Client) TTLResult(ctx context.Context, key string) (time.Duration, err
...
+	// UniversalClient only exposes Cmdable (not StatefulCmdable), so AUTH must be
+	// issued via a pipeline. The pipeline must be Exec'd for the command to actually
+	// be sent over the wire — without it the AUTH is silently queued and dropped.
 	pipeline := c.client.Pipeline()
 	statusCmd := pipeline.AuthACL(ctx, username, password)
+	if _, err := pipeline.Exec(ctx); err != nil {
+		return err
+	}
 	return statusCmd.Err()
 }
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/v9client.go:330]: Same as the v8 case: `Pipeliner` has no `Close()` method, so this wouldn't compile, and a non-Tx pipeline only borrows a pooled connection during `Exec` (returning it right after), so there's nothing to release.

File: common/component/redis/entraid_onconnect_test.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	switch client := c.(v8Client).client.(type) {
+	case *v8.Client:
+		return client.Options().OnConnect, client.Options().Password
+	case *v8.ClusterClient:
+		return client.Options().OnConnect, client.Options().Password
+	default:
+		t.Fatalf("unexpected v8 client type %T", client)
+		return nil, ""
+	}
+}
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/entraid_onconnect_test.go:71]: These are safe as-is: `newV8Client`/`newV9Client` and their failover variants all `return v8Client{...}` / `v9Client{...}` by **value** (the methods use value receivers), so the `c.(v8Client)` / `c.(v9Client)` assertions match the stored concrete type and won't panic. The tests pass in CI for exactly this reason. I'm happy to add pointer cases to the type switch defensively if you'd prefer, but there's no current code path that returns a pointer.

File: common/component/redis/entraid_onconnect_test.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	switch client := c.(v9Client).client.(type) {
+	case *v9.Client:
+		return client.Options().OnConnect, client.Options().Password
+	case *v9.ClusterClient:
+		return client.Options().OnConnect, client.Options().Password
+	default:
+		t.Fatalf("unexpected v9 client type %T", client)
+		return nil, ""
+	}
+}
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/entraid_onconnect_test.go:86]: Same as the v8 helper above — the constructors return `v8Client`/`v9Client` by value, so the value-form type assertion is correct and matches what's stored in the interface.

**Comment by paulstadler-mesh** [comment]
cc @cicoyle @JoshVanL — would one of you be able to take a look at this, or point me to the right reviewer? (I can't add reviewers directly as an external contributor.)

Quick context on why it matters:
- Fixes #3554 — Microsoft Entra ID auth against Azure Managed Redis fails with `WRONGPASS` once the initial token's TTL expires, because new go-redis pool connections replay the token snapshotted at init. This wires an `OnConnect` hook that re-AUTHs each new connection with a freshly-acquired token (complementary to the existing background refresh routine, which only re-AUTHs already-open connections).
- Open since April, CI green, DCO signed.
- We're running this in production on a patched 1.17.3 build, so the fix is validated against real Azure Managed Redis.
- Rebased/merged onto current `main`; conflicts with the new OIDC feature (#4408) are resolved — the `AuthACL` `Exec` fix is now shared with that PR, so the diff is just the `OnConnect` path + EntraID init refactor + tests.

Happy to make any changes you'd like. Thanks!

File: common/component/redis/settings.go [hunk 6]
```diff
@@ -123,6 +133,48 @@ type Settings struct {
...
+// token suitable for use as the AUTH password. It must only be called when UseEntraID is
+// true and after InitEntraIDCredential has succeeded; otherwise it returns an error.
+//
+// This is invoked from the OnConnect callback installed on the underlying go-redis client
+// so that every new pool connection authenticates with a current token, rather than the
+// stale snapshot Password that would otherwise be sent during initial AUTH.
+func (s *Settings) EntraIDFetchAuthArgs(ctx context.Context) (username, password string, err error) {
+	if s.entraIDUsername == "" {
+		return "", "", errors.New("redis client: EntraID username (OID) not initialized; AUTH ACL requires a non-empty username")
+	}
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/settings.go]: Done in 8eeeae27 — renamed to `entraIDFetchAuthArgs` (unexported). It's internal wiring only (OnConnect + refresh routine), and the tests are in-package, so it doesn't need to be part of the public API.

File: common/component/redis/redis.go [hunk 7]
```diff
@@ -211,16 +210,18 @@ func ParseClientFromProperties(properties map[string]string, componentType metad
 		return nil, nil, fmt.Errorf("redis client configuration error: %w", err)
 	}
 
-	// start the token refresh goroutine
+	// start the token refresh goroutine — keeps long-lived connections re-AUTHed via AUTH ACL
+	// before the prior token expires. This is complementary to the OnConnect hook that handles
+	// brand-new pool connections.
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/redis.go]: Good catch — reworded in 8eeeae27. The comment now notes that AUTH is per-connection, so the routine re-authenticates pooled connections as they're exercised rather than every open connection at once.

File: common/component/redis/redis.go [hunk 8]
```diff
@@ -293,45 +292,74 @@ func StartEntraIDTokenRefreshBackgroundRoutine(client RedisClient, username stri
...
+		return nil, fmt.Errorf("redis client configuration error: %w", err)
 	}
 
-	s.Password = token.Token
-
-	// This token has already been validated by EntraID. We use insecure parsing to get the object ID.
+	// Validate the initial token and extract the OID claim used as the Redis ACL username.
+	// We use insecure parsing here because the token was just issued by Entra ID and we are
+	// only inspecting it — Redis itself does the authoritative validation.
 	parsedToken, err := jwt.ParseString(token.Token, jwt.WithVerify(false), jwt.WithValidate(false))
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/redis.go]: Fixed in 8eeeae27 — the comment now says the token is parsed *without* verification/validation (signature and expiry intentionally not checked), and that Redis does the authoritative validation at AUTH.

File: common/component/redis/v8client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/v8client.go:327]: This is now moot: after merging current `main`, we took the upstream `AuthACL` implementation, since the same `Exec` fix landed independently via #4408. The behavioral fix is on `main` and no longer unique to this PR, so a regression test for it belongs with that change rather than here.

File: common/component/redis/entraid_onconnect_test.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+		err := entraIDOnConnectV9(s)(context.Background(), nil)
+		require.Error(t, err)
+	})
+
+	t.Run("v8 surfaces missing-OID as a dial error", func(t *testing.T) {
+		s := &Settings{UseEntraID: true, entraIDTokenCredential: fakeTokenCredential{token: "fresh-token"}}
+		err := entraIDOnConnectV8(s)(context.Background(), nil)
+		require.Error(t, err)
+	})
+}
```
[the remark this answers is not in the document]
Comment (paulstadler-mesh) [reply common/component/redis/entraid_onconnect_test.go:208]: Done in 8eeeae27 — added the v9 missing-OID case (`entraIDOnConnectV9`) alongside the v8 one for parity.

**Comment by paulstadler-mesh** [comment]
@mikeee — you're listed as the reviewer on this one, so tagging you directly. Would you be able to take a look at this stage?

It's in a clean spot now:
- Fixes #3554 — Entra ID auth against Azure Managed Redis fails with `WRONGPASS` once the initial token's TTL expires, because new go-redis pool connections replay the token snapshotted at init. This adds an `OnConnect` hook that re-AUTHs each new connection with a freshly-acquired token (complementary to the existing background refresh routine).
- Up to date with `main`; conflicts with the new OIDC feature (#4408) are resolved.
- All Copilot review threads addressed and resolved (0 open), CI/DCO green.
- We're running this in production on a patched 1.17.3 build, so it's validated against real Azure Managed Redis.

Happy to make any changes you'd like. Thanks!

**Comment by paulstadler-mesh** [comment]
Hi @mikeee @cicoyle @JoshVanL — quick update with some production data on this fix.

We needed it in the meantime, so we've been running this exact code as a pluggable component next to stock daprd 1.17.9 — now serving the Redis state store for ~30 services across our dev/sandbox/prod clusters (Azure Managed Redis + Entra ID, actors included). Zero `WRONGPASS` since rollout, including in dev where we set `maxConnAge: 5m` so every connection re-auths ~12×/hour.

I also have a deterministic test for the token-TTL boundary (rotating stub credential + miniredis password rotation — asserts new connections auth with the refreshed token, and that the pre-fix snapshot behavior fails the same scenario). Happy to push it to this branch — I think it covers the earlier review comment about validating fresh-token behavior end-to-end.

Happy to make any changes needed to get this over the line. Thanks!

**Comment by paulstadler-mesh** [comment]
Pushed the token-TTL boundary test mentioned above (`TestEntraIDTokenRotationAcrossConnections`) — it simulates the token expiring server-side and asserts new pool connections authenticate with the refreshed token via OnConnect, plus a counterfactual showing the pre-fix snapshot-password behavior fails at exactly that boundary. Also related: #4498 fixes a separate pipeliner bug we found in the same area while running this in production.
